binders/kafka-binder/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaPartitionHandler.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2016-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.springframework.cloud.stream.binder.kafka;
+
+import java.lang.reflect.Field;
+import java.util.Map;
+
+import org.springframework.beans.factory.BeanFactory;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.cloud.stream.binder.PartitionKeyExtractorStrategy;
+import org.springframework.cloud.stream.binder.PartitionSelectorStrategy;
+import org.springframework.cloud.stream.binder.ProducerProperties;
+import org.springframework.context.expression.BeanFactoryResolver;
+import org.springframework.expression.EvaluationContext;
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.ReflectionUtils;
+import org.springframework.util.StringUtils;
+
+
+/**
+ * Based(nearly copy) on org.springframework.cloud.stream.binder.PartitionHandler but with partitionCount as parameter.
+ */
+public class KafkaPartitionHandler {
```

Comment:
`PartitionHandler` is a public class that is open for extension. Can you extend from `PartitionHandler` and only add the `determinePartition` method with the extra `partitionCount` parameter? That way, we can avoid duplicating all this code in `KafkaPartitionHandler`.

Reply:
My first try was to extend existing PartitionHandler, but this doesnt't work because of SCS architecture. The necessary KafkaTemplate class (which contains updated partition count) is stored within 'KafkaMessageChannelBinder.ProducerConfigurationMessageHandler' and this class cannot be accessed from existing PartitionHandler. You get a cyclic maven dependency error if you try access it from spring-cloud-stream-core. I see no other possibilty as to add the kafka partition handling code directly to the othe kafka code within 'spring-cloud-stream-binder-kafka'. Maybe I missed an other soloution!?

Reply:
I wasn't proposing you modify `PartitionHandler` in `spring-cloud-stream-core.` Why can't you keep the `KafkaPartitionHandler` in the Kafka binder (as it is right now) and then extend it from `PartitionHandler`. ? You then call the super constructor from the new extension. Later on, in the `handleMessage()` call, you call `determinePartition(message, partitionCount)`, which is only available in `KafkaPartitionHandler.` I don't see how this may introduce a cycle.

Reply:
Ok I got it. My code was not meant to be part of a pull request from the beginning. I will rewrite the KafkaPartitionHandler and will remove the sample application (which was more for my own purpose). Still no idea how to manually test without using my sample program.